compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/DelegationLowering.kt

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.ClassLoweringPass
+import org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext
+import org.jetbrains.kotlin.backend.common.deepCopyWithVariables
+import org.jetbrains.kotlin.backend.common.descriptors.WrappedFieldDescriptor
+import org.jetbrains.kotlin.backend.common.descriptors.synthesizedName
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.descriptors.*
+import org.jetbrains.kotlin.descriptors.annotations.Annotations
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.*
+import org.jetbrains.kotlin.ir.symbols.*
+import org.jetbrains.kotlin.ir.symbols.impl.*
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.types.impl.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.*
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.Variance
+
+internal val propertyDelegationPhase = makeIrFilePhase(
+    ::PropertyDelegationLowering,
+    name = "PropertyDelegation",
+    description = "Construct KProperty instances to be used when accessing delegated properties"
+)
+
+internal class PropertyDelegationLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    private object DECLARATION_ORIGIN_KPROPERTIES_FOR_DELEGATION :
+        IrDeclarationOriginImpl("KPROPERTIES_FOR_DELEGATION", true)
+
+    private val plainJavaClass =
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
```

Comment:
AsmType.REFLECTION

Reply:
Can you clarify, please? Do you mean I should replace `"kotlin.jvm.internal.Reflection"` with `AsmTypes.REFLECTION.replace("/", ".")`?

Reply:
It's more tricky than I though:
JvmClassName.byInternalName(AsmTypes.REFLECTION).fqNameForClassNameWithoutDollars